File: 2021/12/python/main.py
import pathlib

def twice(visited):
  for k, v in visited.items():
    if k.islower() and v > 1:
      return True

  return False

def search(node, visited, part2=False):
  if node not in visited:
    visited[node] = 1
  else:
    visited[node] += 1

  if node == "end":
    return [["end"]]

  paths = []

  for connection in cave[node]:
    if connection == "start":
      continue
    if connection not in visited or connection.isupper() or (part2 and not twice(visited)):
      result = search(connection, visited.copy(), part2)
      if result:
        paths += [
          [node] + path for path in result
        ]

  return paths
        

cave = {}
path = "../input.txt"
puzzle_input = pathlib.Path(path).read_text().strip()

for line in puzzle_input.splitlines():
  a, b = line.strip().split("-")

  if a not in cave:
    cave[a] = set()
  if b not in cave:
    cave[b] = set()

  cave[a].add(b)
  cave[b].add(a)

paths = search("start", {"start": 0})
print("Part 1:", len(paths))

paths = search("start", {"start": 0}, True)
print("Part 2:", len(paths))

File: 2021/12/rust/src/main.rs
use std::collections::{HashMap, HashSet};
use std::error::Error;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn visited_twice(visted: &HashMap<&str, u8>) -> bool {
    for (k, v) in visted.iter() {
        if *k == k.to_ascii_lowercase() && *v > 1 {
            return true;
        }
    }

    false
}

fn search<'a>(
    caves: &'a HashMap<String, HashSet<String>>,
    node: &'a str,
    visited: &mut HashMap<&'a str, u8>,
    part2: bool,
) -> Vec<Vec<&'a str>> {
    *visited.entry(node).or_default() += 1;

    let mut paths: Vec<Vec<&str>> = Vec::new();

    if node == "end" {
        paths.push(vec!["end"]);
        return paths;
    }

    for connection in caves.get(node).unwrap() {
        if connection == "start" {
            continue;
        }

        if !visited.contains_key(connection.as_str())
            || *connection == connection.to_ascii_uppercase()
            || (part2 && !visited_twice(&visited))
        {
            let results = search(caves, connection, &mut visited.clone(), part2);
            paths.extend(
                results
                    .iter()
                    .map(|path| [vec![node], path.clone()].concat())
                    .collect::<Vec<Vec<&str>>>(),
            );
        }
    }

    paths
}

fn main() -> Result<(), Box<dyn Error>> {
    let file = File::open("../input.txt")?;
    let reader = BufReader::new(file);

    let mut caves: HashMap<String, HashSet<String>> = HashMap::new();

    for line in reader.lines() {
        let line = line?;
        let (a, b) = line.trim().split_once("-").unwrap();
        for (x, y) in [(a, b), (b, a)] {
            caves
                .entry(x.to_string())
                .or_default()
                .insert(y.to_string());
        }
    }

    let mut visited: HashMap<&str, u8> = HashMap::new();
    let paths = search(&caves, "start", &mut visited, false);
    println!("Part 1: {}", paths.len());

    let mut visited: HashMap<&str, u8> = HashMap::new();
    let paths = search(&caves, "start", &mut visited, true);
    println!("Part 2: {}", paths.len());

    Ok(())
}
